### PR TITLE
Unicode 9 updates

### PR DESCRIPTION
**Pull request by Keno**
I believe the only substantiative changes to the code required are the new rules in TR29, but please do check the Unicode change log (http://www.unicode.org/versions/Unicode9.0.0/) for other changes that may affect this library.

### CODE REVIEW

File: utf8proc.c [hunk 1]
```diff
@@ -233,33 +233,56 @@ UTF8PROC_DLLEXPORT const utf8proc_property_t *utf8proc_get_property(utf8proc_int
...
 }
 
-/* return whether there is a grapheme break between codepoints c1 and c2 */
+/* return whether there is a grapheme break between codepoints c1 and c2.
+
+  CAVEATS:
+   Please note that evaluation of GB10 (grapheme breaks between emoji zwj sequences)
+   and GB 12/13 (regional indicator code points) require knowledge of previous characters
+   and are thus not handled by this function. This may result in an incorrect break before
+   and E_Modifier class codepoint and an incorrectly missing break between two
```
Comment (tkelman) [inline utf8proc.c]: an E_modifier?

**Comment by tkelman** [comment]
This almost certainly needs version number bumps

File: utf8proc.c [hunk 2]
```diff
@@ -389,7 +412,21 @@ UTF8PROC_DLLEXPORT utf8proc_ssize_t utf8proc_decompose_char(utf8proc_int32_t uc,
...
+    if (*last_boundclass == tbc && tbc == UTF8PROC_BOUNDCLASS_REGIONAL_INDICATOR)
+      *last_boundclass = UTF8PROC_BOUNDCLASS_OTHER;
+    // Special support for GB10. Fold any EXTEND codepoints into the previous
+    // boundclass if we're dealing with an emoji base boundclass.
+    else if ((*last_boundclass == UTF8PROC_BOUNDCLASS_E_BASE      ||
+              *last_boundclass == UTF8PROC_BOUNDCLASS_E_BASE_GAZ) &&
+             tbc == UTF8PROC_BOUNDCLASS_EXTEND)
+      *last_boundclass = UTF8PROC_BOUNDCLASS_E_BASE;
+    else
+      *last_boundclass = tbc;
```
Comment (stevengj) [inline utf8proc.c]: Would be nice to export this logic somehow so that we can use it in the Julia grapheme iterator.

Comment (Keno) [inline utf8proc.c]: Mhm, I believe just adding an extra out parameter to give you the override class should be fine.

**Comment by Keno** [comment]
Ok, I've updated the API to expose the state override. Since that's an ABI incompatible change, I've bumped the MAJOR version accordingly. Please review.

**Comment by Keno** [comment]
We might want to do this at the same time as #68 (which may also require a rebase for this). cc @benibela

File: Makefile [hunk 3]
```diff
@@ -19,9 +19,9 @@ UCFLAGS = $(CFLAGS) $(PICFLAG) $(C99FLAG) $(WCFLAGS) -DUTF8PROC_EXPORTS
 # not API compatibility: MAJOR should be incremented whenever *binary*
 # compatibility is broken, even if the API is backward-compatible
 # Be sure to also update these in MANIFEST and CMakeLists.txt!
```
Comment (stevengj) [inline Makefile]: Did you forget to update CMakeLists.txt too?

Comment (Keno) [inline Makefile]: Yes

Comment (stevengj) [inline Makefile]: Note that you also need to update `utf8proc.h` for the API version change.

**Comment by stevengj** [comment]
Maybe better to do this first, then update #68?

File: utf8proc.h [hunk 4]
```diff
@@ -520,8 +520,19 @@ UTF8PROC_DLLEXPORT utf8proc_ssize_t utf8proc_reencode(utf8proc_int32_t *buffer,
...
+ *
+ * @param state Beginning with Version 29 (Unicode 9.0.0), this algorithm requires
+ *              state to break graphemes. This state can be passed in as a pointer
+ *              in the `state` argument and should initially be set to 0. If the
+ *              state is not passed in (i.e. a null pointer is passed), UAX#29 rules
+ *              GB10/12/13 which require this state will not be applied, essentially
+ *              matching the rules in Unicode 8.0.0.
+ *
+ * @warning If the state parameter is used, `utf8proc_grapheme_break` must be called
+ *          IN ORDER on ALL potentital breaks in a string.
```
Comment (stevengj) [inline utf8proc.h]: potential

File: utf8proc.c [hunk 5]
```diff
@@ -273,19 +282,38 @@ static utf8proc_bool grapheme_break(int lbc, int tbc) {
...
 }
 
-/* return whether there is a grapheme break between codepoints c1 and c2.
+static utf8proc_bool grapheme_break_extended(int lbc, int tbc, utf8proc_int32_t *state)
+{
+  int lbc_override = lbc;
+  if (state && *state != UTF8PROC_BOUNDCLASS_START)
+    lbc_override = *state;
+  utf8proc_bool break_permitted = grapheme_break_simple(lbc, tbc);
+  if (state) {
```
Comment (stevengj) [inline utf8proc.c]: Is it really necessary to have general state like this?  Wouldn't it be enough just to know the previous codepoint before `lbc`?

Comment (Keno) [inline utf8proc.c]: No, there may be an infinite number of EXTEND characters between E_Base and E_Modifier.

**Comment by Keno** [comment]
Do we need to hold off on doing the version bump if we want to do #68 right after?

**Comment by tkelman** [comment]
I think we can do a version bump for this (definitely needed), and if we want to merge further big changes we don't need to change the version again until we do the next release.

**Comment by Keno** [comment]
Ok.

**Comment by Keno** [comment]
Good to merge?

**Comment by benibela** [comment]
There is a warning about an unused variable: lbc_override

**Comment by Keno** [comment]
Oops, thanks.

**Comment by Keno** [comment]
I also realized I made this static. Will fix.

File: utf8proc.c [hunk 6]
```diff
@@ -233,36 +233,87 @@ UTF8PROC_DLLEXPORT const utf8proc_property_t *utf8proc_get_property(utf8proc_int
...
 
-/* return whether there is a grapheme break between codepoints c1 and c2 */
-UTF8PROC_DLLEXPORT utf8proc_bool utf8proc_grapheme_break(utf8proc_int32_t c1, utf8proc_int32_t c2) {
-  return grapheme_break(utf8proc_get_property(c1)->boundclass,
-                        utf8proc_get_property(c2)->boundclass);
+static utf8proc_bool grapheme_break_extended(int lbc, int tbc, utf8proc_int32_t *state)
+{
+  int lbc_override = lbc;
+  if (state && *state != UTF8PROC_BOUNDCLASS_START)
+    lbc_override = *state;
```
Comment (stevengj) [inline utf8proc.c]: @Keno, I just noticed that `lbc_override` seems to never be used.  Did you mean to pass `lbc_override` to `grapheme_break_simple`?

Comment (Keno) [reply utf8proc.c]: It's been a while, but yes, that looks right, esp by looking at the first commit in this PR.

Comment (Keno) [reply utf8proc.c]: We already seem to have found that bug though: https://github.com/JuliaLang/utf8proc/blob/master/utf8proc.c#L290

Comment (Keno) [reply utf8proc.c]: https://github.com/JuliaLang/utf8proc/pull/77

**Pull request merged**